### ISSUE TITLE
Design an alternative to the dynamic scope bug

### ISSUE THREAD

**Issue by chameco** [GaloisInc/saw-script#1646]
Currently top-level `let` bindings in SAWScript behave a little strangely. Consider the following:

```
let COND = true;

let f x y = if COND then x else y;

print (f "foo" "bar");

let COND = false;

print (f "foo" "bar");
```

This prints "foo" and then "bar" rather than the expected "foo" and then "foo".

While strange, this behavior is sometimes useful (and indeed is used in the s2n and BLST proofs, in order to dry-run sections of the proof during development), so we should provide an alternative. We should consider the remote API in this discussion - fundamentally, I think the usefulness of this "feature" stems from unclear boundaries between the interface to SAW-the-system (imperative, works with proof commands and global state updates) and SAWScript (a collection of DSLs to write specs / proof tactics). Clarifying the boundary between these "languages" would be broadly useful - Coq does something similar here with its "vernacular" command language built atop expression/tactic languages.

**Comment by robdockins** [comment]
One obvious solution to this is to implement actual ML-style `ref` cells with more explicit syntax for reading and updating them.

**Comment by brianhuffman** [comment]
We should prioritize this. The dynamic scoping bug means that saw-script is not type safe, because we can redefine the variable in question to a new value with a different type:
```
sawscript> let x = "a"
sawscript> let f() = print (str_concat "x=" x)
sawscript> f()
[21:20:13.464] x=a
sawscript> let x = "b"
sawscript> f()
[21:20:18.536] x=b
sawscript> let x = 5
sawscript> f()

fromValue String
CallStack (from HasCallStack):
  error, called at saw-central/src/SAWCentral/Value.hs:1226:19 in saw-1.3.0.99-inplace-saw-central:SAWCentral.Value
```

**Comment by sauclovian-g** [comment]
Hmm, good point, that definitely shouldn't be allowed.

I wonder if it's adequate to paper over this temporarily by requiring rebindings of existing variables to have the same type. Because otherwise to provide an alternative, we need to add some kind of assignment, and while I'm not in any way opposed to that it's a can of worms that I don't think we're ready to open just yet.

**Comment by sauclovian-g** [comment]
Ok, here's a proposal for an alternative mechanism that is kind of hacky but avoids major language changes:

1. Add a new reserved word `rebindable`. Allow this on variable bindings, as in `rebindable let x = 3` or `rebindable x <- foo`. (Or maybe `let rebindable`.)

2. Variables marked `rebindable` can be rebound later with the current update behavior. Any such rebinding must have the same type.

3. Function arguments can't be `rebindable`.

4. Warn when any other rebind occurs (that is, a new binding of the same variable in the same scope). Maybe also warn about shadowing (new bindings in nested scopes).

5. Internally (during execution) look up and resolve variable names that aren't function arguments when closing over the body of a function or lambda, except for those declared `rebindable`. This will prevent seeing new bindings of non-`rebindable` variables.

Note that currently the behavior only works for bindings at the syntactic top level. (Just being inside a do-block inhibits it.) We could continue that restriction by explicitly only allowing `rebindable` at the syntactic top level, or allow it to happen more generally. I don't have a strong opinion on this point. (The special cases in execution that currently make this distinction possible should be removed.)

**Comment by RyanGlScott** [comment]
I don't have a particularly strong objection to this plan, although I do agree that it is hacky. In that sense, perhaps it's for the best that you'd have to type a particularly long keyword like "`rebindable`" to use this feature, as we don't people to make use of this unless they _really_ want it.

In a future SAWScript 2.0, I would like to think about a more significant change that introduces a notion of proper mutable references with assignment instead. (That isn't something that we have to do right now, of course.)

**Comment by sauclovian-g** [comment]
Yes, I agree, if we're making large changes we want ref cells of some kind; the immediate problem is how to deal with the current unsafe situation without making large language changes.

**Comment by sauclovian-g** [comment]
Ok, here's what got implemented, more or less as above:

1. Add a new reserved word `rebindable`. Allow this on variable bindings. The concrete syntax is `let rebindable x = 3`.

2. `rebindable` is permitted only on `let`. It is not permitted on `<-` or `rec`. (Either of these restrictions could be lifted at little cost.)

3. Only variables in the top level scope (the syntactic top level) can be `rebindable`. Locals and function arguments cannot be. This restriction is not so easily lifted.

4. A variable bound with `let rebindable` can be updated by binding it again with another `let rebindable`. The type must be the same. All references to a `rebindable` variable resolve to the most recently bound value. `rebindable` variables are effectively mutable globals.

5. Polymorphic values can be bound as `rebindable` but cannot be rebound. This restriction could be lifted if we felt like implementing the necessary type logic and satisfying ourselves it's safe.

6. Non-`rebindable` variables behave normally: they are immutable and get captured when you close over functions and do-blocks. You can bind new variables with the same name (though this will, for the time being, always trigger a warning) but they're new variables and don't affect the old ones.

7. You can rebind a `rebindable` variable without using `rebindable`; this hides the `rebindable` variable with an ordinary one, either temporarily if in a nested scope or permanently if at the top level.

The commits noted above implement roughly half of this; the rest is coming shortly.
